myhoard/util.py

```diff
@@ -469,3 +470,49 @@ def track_rate(*, current, last_recorded, last_recorded_time, metric_name, min_i
             return_value = current
             return_time = now
     return return_value, return_time
+
+
+class RateTracker(threading.Thread):
+    """Monitors rates of change, somewhat similarly to track_rate(), but capable of updating
+    the statsd gauge even when there is no (or little) activity.
+
+    frequency: period in seconds between recalculation of the metric
+    window: minimum period of time over which the rate is calculated.
+    """
+
+    def __init__(self, *, stats, metric_name, frequency: float = 10, window: float = 30) -> None:
+        super().__init__(daemon=True)
+        self.stats = stats
+        self.metric_name = metric_name
+        self.frequency: float = frequency
+        self.window: float = window
+        self._current_rate: Optional[int] = None
+        self._previous_values_lock = threading.Lock()
+        self._previous_values: List[Tuple[float, float]] = []
+        self._running = True
+
+    def increment(self, value: float) -> None:
+        now = time.monotonic()
+        with self._previous_values_lock:
+            self._previous_values.append((now, value))
+
+            # delete old values until the oldest value is "just" over 'window' seconds ago
+            while len(self._previous_values) > 1 and self._previous_values[1][0] + self.window < now:
+                self._previous_values.pop(0)
```

Comment:
I'm worried about the memory usage (total amount and general overhead from list manipulation) if the network speed is large and/or packets are small and/or the window is large.

It could be safer to do an exponential averaging :
```python
time_constant_secs = 4
delta_time_secs = now - last_update
smooth_value = value - math.exp(-delta_time_secs/time_constant_secs) * (value - smooth_value)
```
A time constant of 4 takes 59sec to go from 1MB/sec to ~0, 86sec to go from 1GB/sec to ~0, 114sec to go from 1TB/sec to ~0.

But if we want a perfectly clean graph (that goes from max to zero between two ticks of the frequency, after rounding of the value), we can pick it from the update frequency and the maximum expected transfer rate:

```python
time_constant_secs = (frequency_secs - math.log(2)) / math.log(max_speed_bytes)
```
For 30sec and 1GB/sec, we get a time constant of 1.4.

Reply:
Just a brief reply/thought: one possibility to ensure an upper bound on memory use is to round each timestamp by (say) window/10. If the new value has the same adjusted timestamp as the latest value, we combine them.

This would result in at most 10 (or 11) items in the list, and would still provide a fairly "snappy" response to variations in transfer rates.